src/chip/drivers/ccm_118x.rs

```diff
@@ -0,0 +1,1952 @@
+        Lpuart0506: 29,
+        Lpuart0708: 30,
+        Lpuart0910: 31,
+        Lpuart1112: 32,
+        Lpi2c0102: 33,
+        Lpi2c0304: 34,
+        Lpi2c0506: 35,
+        Lpspi0102: 36,
+        Lpspi0304: 37,
+        Lpspi0506: 38,
+        I3c1: 39,
+        I3c2: 40,
+        Usdhc1: 41,
+        Usdhc2: 42,
+        Semc: 43,
+        Adc1: 44,
+        Adc2: 45,
+        Acmp: 46,
+        Ecat: 47,
+        Enet: 48,
+        Tmr1588: 49,
+        Netc: 50,
+        Mac0: 51,
+        Mac1: 52,
+        Mac2: 53,
+        Mac3: 54,
+        Mac4: 55,
+        Serdes0: 56,
+        Serdes1: 57,
+        Serdes2: 58,
+        Serdes01g: 59,
+        Serdes11g: 60,
+        Serdes21g: 61,
+        Xcelbusx: 62,
+        Xriocu4: 63,
+        Mctrl: 64,
+        Sai1: 65,
+        Sai2: 66,
+        Sai3: 67,
+        Sai4: 68,
+        Spdif: 69,
+        Asrc: 70,
+        Mic: 71,
+        Cko1: 72,
+        Cko2: 73,
+    );
+
+    /// Configure a clock root
+    ///
+    /// Divider ([`div`]) must be reasonable, follow table 138 in 20.4.3 chapter regarding legal frequencies.
+    pub fn configure<T: Clockroot>(ccm: &mut CCM, clockroot: T, div: u8) {
+        write_reg!(
+            ral::ccm::clockroot, &ccm.CLOCK_ROOT[T::INDEX], CLOCK_ROOT_CONTROL,
+            MUX: clockroot.mux(),
+            DIV: div.saturating_sub(1) as u32, // actualDIV == DIV+1
+        );
+    }
+
+    /// Inspect clock root registers and print out values via `defmt`
+    pub fn inspect<T: Clockroot>(ccm: &mut CCM) {
+        let val = read_reg!(
+            ral::ccm::clockroot,
+            &ccm.CLOCK_ROOT[T::INDEX],
+            CLOCK_ROOT_CONTROL
+        );
+        defmt::info!(
+            "CLOCK_ROOT[{} / {}].CLOCK_ROOT_CONTROL: {:#x}",
+            T::INDEX,
+            T::NAME,
+            val
+        );
+        let val = read_reg!(
+            ral::ccm::clockroot,
+            &ccm.CLOCK_ROOT[T::INDEX],
+            CLOCK_ROOT_STATUS0
+        );
+        defmt::info!(
+            "CLOCK_ROOT[{} / {}].CLOCK_ROOT_STATUS0: {:#x}",
+            T::INDEX,
+            T::NAME,
+            val
+        );
+    }
+}
+
+/// Initialize and configure clocks
+///
+/// - switch M33 core temporarily to RC 400MHz
+/// - configure DCDC 1p0 power rail to serve 1.1V
+/// - initialize external 24MHz oscillator
+/// - initialize SysPLL3
+/// - switch M33 core to SysPLL3 with div: 2 (240 MHz)
+/// - switch FlexSPI1 to SysPll3Pfd0 with div: 5 (~133MHz)
+///
+/// Use [`clockroot::configure`] function in order to configure more clockroots
+pub fn init(
+    anadig_osc: &mut ANADIG_OSC,
+    anadig_pll: &mut ANADIG_PLL,
+    anadig_pmu: &mut ANADIG_PMU,
+    ccm: &mut CCM,
+    dcdc: &mut DCDC,
+    phy_ldo: &mut PHY_LDO,
+) {
+    // Unclear if it is necessary
+    // - C codegen from clock tool does this
+    // - BootROM does this when running from flash
+    // Doing the same thing for the sake of completeness.
+    defmt::debug!("init osc rc 400m");
```

Comment:
#183 introduced `defmt` as an _optional_ feature. However, this code assumes the `defmt` feature is enabled. We'll fail to build when the `defmt` feature isn't enabled, like

```
cargo build --features=imxrt-ral/imxrt1189_cm33,imxrt1180 --package=imxrt-hal
```

Could you rebase onto #186, which adds CI coverage for this build? If we intend on keeping `defmt` optional, we shouldn't unconditionally emit these logs.

Reply:
Fixed